### ISSUE TITLE
Improve Chinese Segmentation & Word Sense Disambiguation with Transformer-based ML

### ISSUE THREAD

**Issue by oliverzh2000** [oliverzh2000/wen-reader#2]
## Summary

Replace the current rule-based Chinese word segmentation and add ML-based word sense disambiguation (WSD) to dramatically improve dictionary lookup quality for literary/classical Chinese texts.

## Current State

| Feature | Current Approach | Accuracy | Issue(s) | 
|---------|-----------------|----------|-------| 
| Segmentation | DP algo in `Services/SegmentationService.swift` | TODO | Struggles with heavily with ambiguity and context | 
| WSD | None (first sense) | N/A | |

### 1. Segmentation: CKIP BERT Transformer

__Model__: [ckiplab/bert-base-chinese-ws](https://huggingface.co/ckiplab/bert-base-chinese-ws)

- Pre-trained Chinese BERT fine-tuned on word segmentation
- \~110M parameters
- No training needed - just export to CoreML

__Test Results:__

### 2. WSD: BEM-Style Bi-Encoder trained on MiCLS

__Problem__: No pre-trained Chinese WSD model exists publicly.

__Solution__: Fine-tune Chinese BERT using BEM (Bi-Encoder Model) approach on MiCLS dataset.

__Architecture__:

```javascript
context: "他今天看起来特别[TGT]精神[/TGT]"
         ↓ (BERT encoder)
    context_embedding

candidate_sense_1: "vigor; vitality"
         ↓ (BERT encoder)
    sense_embedding_1 

candidate_sense_2: "spirit / mind / consciousness"
         ↓ (BERT encoder)
    sense_embedding_2

Cosine similarity between context_embedding and (sense_embedding_1, ..., sense_embedding_n)
         ↓
    Pick highest scoring sense
```

__Dataset__: [MiCLS](https://github.com/COOLPKU/MorBERT) (80k examples)

- Largest public Chinese WSD dataset
- From ACL 2024 MorBERT paper

__Expected Results__: | Approach | Accuracy | |----------|----------| | Current (none) | 0% (always first sense) | | Baseline bge-m3 | 71% | | __BEM + MiCLS__ | ~85% | | MorBERT SOTA | 92% |

### Segmentation (CKIP BERT)

- [x] Test CKIP BERT model in Python
- [x] Verify accuracy on literary Chinese
- [x] Export model to CoreML
- [ ] Create Swift `SegmentationService` with CoreML
- [ ] Integrate into reader long-press flow
- [ ] Test on device performance

### Word Sense Disambiguation (BEM)

- [x] Research WSD approaches (BEM, EWISER, MorBERT)
- [x] Test baseline embedding approach (71%)-
- [x] Identify training data (MiCLS dataset)
- [x] Document training plan (`ml/docs/chinese_wsd_training_plan.md`)
- [x] Download MiCLS training data
- [x] Write BEM training script
- [x] Train model (~2-4 hrs on GPU)
- [x] Evaluate on MiCLS test set
- [x] Export to CoreML
- [x] Build sql with machine translated chinese senses
  - [x] Find a good translation model to run locally (API too expensive)
  - [x] Build sql
- [ ] Create Swift `WSDService` with CoreML
- [ ] Integrate into dictionary popup
- [ ] Test on device performance

### Performance & Model Size Optimization
- [ ] Quantize models - need to keep accuracy drop within reasonable limit
- [ ] Distill models if accuracy still an issue
- [ ] Merge cws and wsd vocab.txt together - only difference is the [TGT] and [/TGT] tokens at end of wsd_vocab.txt
- [ ] 

### Papers

- [BEM (ACL 2020)](https://aclanthology.org/2020.acl-main.95/) - Bi-Encoder WSD approach
- [MorBERT (ACL 2024)](https://aclanthology.org/2024.acl-long.819/) - Chinese WSD SOTA + MiCLS dataset

### Data

- MiCLS Training: [](https://pan.baidu.com/s/1Uw63Bo10fn7jeef8tUThiw?pwd=rxe6)<https://pan.baidu.com/s/1Uw63Bo10fn7jeef8tUThiw?pwd=rxe6>
- MiCLS Val/Test: [](https://github.com/COOLPKU/MorBERT/tree/main/data)<https://github.com/COOLPKU/MorBERT/tree/main/data>

### Models

- CKIP Segmentation: [](https://huggingface.co/ckiplab/bert-base-chinese-ws)<https://huggingface.co/ckiplab/bert-base-chinese-ws>
- Chinese BERT WWM: [](https://huggingface.co/hfl/chinese-bert-wwm-ext)<https://huggingface.co/hfl/chinese-bert-wwm-ext>

## Device Requirements

| Model | Size | Latency (est) | 
|-------|------|---------------| 
| CKIP Segmentation | ~200MB | ~50ms/sentence |
| WSD Context Encoder | ~200MB | ~30ms/word | | Gloss Embeddings | ~50MB | Pre-computed |

**Comment by oliverzh2000** [comment]
## ML Summary: CWS Span Scorer + WSD Bi-encoder

### Overview

Two on-device CoreML models for context-aware Chinese NLP in the reader:

1. **CWS Span Scorer** — given a sentence, scores candidate dictionary word spans and finds the optimal segmentation via dynamic programming
2. **WSD Bi-encoder** — given a word in context, ranks CC-CEDICT sense clusters by cosine similarity to select the correct dictionary definition

Both run on-device (iPhone/iPad) with int8 per-block quantized weights, fixed input shapes, targeting iOS 18+ ANE.

---

### CWS: Span Scorer

#### Architecture

- **Encoder:** ELECTRA-small (12 layers, 256 hidden, 4 heads, ~12M params) — `hfl/chinese-electra-180g-small-discriminator`
- **Span scoring head:** For each candidate span (i, j), concatenates `[H[start] ; H[end] ; width_embedding(len)]` → MLP (576→256→ReLU→1) → scalar score
- **Decoding:** Viterbi-style DP over all CEDICT spans to find the best non-overlapping cover
- **On-device split:** Encoder exported as CoreML (one forward pass per sentence), span head runs as raw Swift matrix math (~410K params, two matmuls + ReLU). Width embeddings are a 19×64 lookup table hardcoded in Swift.

#### Why span scoring over traditional CWS (B/I tagging)?

Traditional CWS predicts per-character boundary labels, then post-processes to get dictionary-compatible words. Span scoring is a more direct formulation: enumerate all CEDICT entries via trie lookup, score them in context, pick the best cover via DP. Output is CEDICT-constrained by construction — no OOV tokens, no sub-splitting heuristics.

#### Training data

~104k sentences, ~914k span-scorer training examples from:

| Source | Sentences | Annotator |
|---|---|---|
| Ebook corpus (14 books) | ~7k | Claude Sonnet (thinking) |
| ICWB2 auto-matched (greedy=gold) | ~28k | Auto-labeled |
| ICWB2 disagreements | ~31k | Claude Opus/Sonnet (thinking) |
| Wiki + OpenSubtitles (ambiguity-selected) | ~38k | Claude Opus (thinking) |

**Data pipeline:**
1. Scan corpus with CEDICT trie to find "conflict points" — positions where 2+ multi-char dictionary words overlap
2. Score sentences by ambiguity density + CEDICT coverage (ensuring broad vocabulary representation)
3. Present to LLM: full sentence + all candidate words at each position + pinyin/definitions → LLM picks the correct segmentation
4. Convert to span training format: gold span = positive, other CEDICT spans at same position = negatives

**Key design decisions:**
- Index-based annotation (LLM picks word indices, not reproducing segmented text) — 100% structural validity vs ~55% for text-reproduction prompts
- Coverage-aware selection: pure ambiguity ranking over-represents common conflicts; coverage bonus ensures rare dictionary words appear in training
- Full-sentence labeling (not just ambiguous positions) to handle cascading ambiguities

#### Training

- Loss: cross-entropy over candidate spans at each position (variable candidate set per position)
- Discriminative learning rates: encoder 1e-4, head 3e-4 (AdamW)
- 4 epochs, batch size 128
- Full fine-tuning (encoder + head end-to-end)

#### Eval (97 hand-curated test cases)

| Model | Perfect | Oversplit | Passed | Wrong |
|---|---|---|---|---|
| Greedy longest match | 74 | 0 | 74/97 | 23 |
| **Electra-small (deployed, int8)** | 83 | 9 | **92/97** | 5 |
| Electra-base (int8) | 88 | 5 | 93/97 | 4 |

Int8 per-block quantization: zero degradation for base, negligible for small.

#### CoreML export

- Fixed input shape: seq_length=32
- int8 per-block (block_size=32) weight quantization
- iOS 18 deployment target
- Bundle size: 13 MB (encoder .mlpackage) + 584 KB (span head weights)

---

### WSD: Bi-encoder

#### Architecture

- **Encoder:** GTE-small-zh distilled from GTE-base-zh (~30M params)
- **Task:** Encode context sentence (with ★target word★ markers) and candidate sense labels, rank by cosine similarity
- **Sense labels:** Chinese translations of CC-CEDICT English definitions, concatenated per sense cluster with "；"

#### Training data

183k positive examples from:

| Source | Examples | Description |
|---|---|---|
| Ebook LLM-annotated | 102k | Real ebook text segmented by CWS model, LLM picks correct sense cluster (Sonnet batch) |
| V1 LLM-generated + MiCLS | 81k | Balanced per-sense examples (crafted sentences + MiCLS corpus mapped to CEDICT) |

**Sense clustering:**
- CC-CEDICT entries with related senses merged using LLM-assisted clustering
- Chinese sense labels translated from English definitions via LLM (cached in `translation_cache.json`)
- 10,092 polysemous words, 23,579 total sense clusters

**Key insight — merging data sources:** V1 (balanced, LLM-generated) provides strong rare-sense anchors. V2 (natural ebook distribution) provides domain matching and realism. Combined: 88.7% top-1 accuracy vs 79.1% (v1 alone) or 74.0% (v2 alone).

#### Training

- Loss: grouped cross-entropy over cosine similarities with temperature τ=0.1
- Contexts grouped by word in each batch — sense embeddings encoded once, reused across all contexts for that word
- LR: 2e-5, 5 epochs, warmup 10%, cosine decay
- Distillation: base→small teacher-student (improves small from 64.9% to 80.8%)

#### Eval (239 hand-curated disambiguation examples)

| Model | TOP-1 | TOP-3 | MRR |
|---|---|---|---|
| **GTE-base (best accuracy)** | 88.7% | 99.6% | 0.936 |
| **Distilled small (deployed, int8)** | 80.8% | 98.3% | 0.889 |

Embedding quantization (int8 per-vector scalar) is completely lossless — unit-normalized vectors with 512 dims have huge margin.

#### CoreML export

- Fixed input shape: seq_length=16
- int8 per-block (block_size=32) weight quantization
- Sense embeddings stored as int8 in CEDICT SQLite database
- Bundle size: 30 MB (encoder .mlpackage)

---

### Deployment Summary

| Component | Model | Params | Bundle | Accuracy |
|---|---|---|---|---|
| CWS encoder | Electra-small-v1 int8 | 12M | 13 MB | 92/97 hard cases |
| CWS span head | MLP (raw weights) | 410K | 584 KB | — |
| WSD encoder | Distilled GTE-small int8 | ~30M | 30 MB | 80.8% top-1, 98.3% top-3 |
| CEDICT + sense embeddings | SQLite (dict + int8 embs) | — | 50 MB | — |
| Vocabs | tokenizer vocabs | — | 180 KB | — |
| **ML total** | | | **~44 MB** | |

Both models use fixed input shapes (no enumerated shapes), enabling ANE execution and eliminating the ~2.6GB buffer pre-allocation issue seen with enumerated shapes on iPhone 13 Pro.

**Runtime performance (iPhone 13 Pro):** 5–40ms per inference for either model.

---

### Tools & Infrastructure

- Training: PyTorch + HuggingFace Transformers + SentenceTransformers
- LLM annotation: Claude Opus/Sonnet, DeepSeek, OpenAI via API for experimentation
- Export: coremltools with custom fp16 overflow patch
- Eval: `eval_coreml.py` runs exported .mlpackage files on macOS to verify end-to-end accuracy post-quantization
- Dictionary: CC-CEDICT trie (simplified forms), sense clusters in `entries_after_merging.json`
